### ISSUE TITLE
The last backup information returned is different from actual. (similar case reported on #5477 earlier)

### ISSUE THREAD

**Issue by bmwluk2012** [sqlcollaborative/dbatools#6730]
Background of the case, the following command is used.
$allservers = Get-DbaCmsRegServer -SqlInstance carmd-iv-sqlmgt -Group PROD
$allservers | Get-DbaLastBackup  | 
              Select-Object *, @{L='CollectDate';E={Get-Date -Format 'MM/dd/yyyy HH:mm:ss'}} |
              ConvertTo-DbaDataTable | 
              Write-DbaDataTable -SqlInstance carmd-iv-sqlmgt -Database DBA_Monitor -Table PS_LastBackupHistory -AutoCreateTable

### Environmental information

PowerShell Version        : 5.1.14393.3471
dbatools latest installed : 1.0.115
Culture of OS             : en-US

#### SQL Server:

```sql
/* REPLACE WITH output of @@VERSION */
```
![image](https://user-images.githubusercontent.com/36716515/89810837-8bc36280-daf2-11ea-9bd1-bc283c808c6c.png)

```sql
/* REPLACE WITH output of @@LANGUAGE */
```
us_english
#### Host used

- [ ] powershell.exe
- [ ] ISE
- [ ] VS Code
- [X] Other (please specify)
SQL Agent job running the following command.
powershell.exe -ExecutionPolicy Bypass -File D:\dbatools_script\Job_LastBackupHistory\LastBackupHistory.ps1
If anything other than powershell.exe was used, please confirm that you can duplicate the issue with powershell.exe

- [ ] Still buggy in powershell.exe

#### Errors Received
Shown in the log file.
WARNING: [22:10:03][Get-DbaBackupHistory] Found backups from multiple recovery 
forks for MISBMM64 on carmd-iv-sql171, this may affect your results

PSMessageDetails      : 
Exception             : System.Management.Automation.RuntimeException: Unable to find type 
                        [Microsoft.PowerShell.Commands.PowerShellGet.Telemetry].
                           at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, 
                        Exception exception)
                           at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
                           at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
                           at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
TargetObject          : Microsoft.PowerShell.Commands.PowerShellGet.Telemetry
CategoryInfo          : InvalidOperation: (Microsoft.Power...llGet.Telemetry:TypeName) [], RuntimeException
FullyQualifiedErrorId : TypeNotFound
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, C:\Program Files\WindowsPowerShell\Modules\PowerShellGet\1.0.0.1\PSModule.psm1: line 678
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}

#### Steps to Reproduce

powershell.exe -ExecutionPolicy Bypass -File D:\dbatools_script\Job_LastBackupHistory\LastBackupHistory.ps1

#### Expected Behavior

I am expected the latest backup information captured.

#### Actual Behavior

The result is false, the backup date is not the actual value.

**Comment by andreasjordan** [comment]
The code looks good, so there is maybe an edge case in the data. Can you provide the information that Get-DbaDbBackupHistory returns for that database?
Or can you reproduce this on a clean instance with a series of Backup-DbaDatabase commands that we can use to build that situation in our labs?

**Comment by bmwluk2012** [comment]
Hi Sqlcollaborative/Dbatools,

Kindly find the screenshot from the table below which is the result wrote
into the table,
The date highlighted are not the true value of the DB backup corresponding
to the DB 'MISBMM64'.  The actual backup date regarding the DB is shown in
the screenshot below which is captured from SSMS via connected to the
corresponding SQL Instance.
[image: image.png]
Actually the DB backup is run on schedule (i.e. daily on DIFF; weekly on
FULL) and job run successfully; this is why I do not understand why
"Get-DbaDatabase" did not return the true value of backup date for this
specific database.   Frankly speaking, I rely on the record returned and
scan for any DB is missing backup currently for taking appropriate action.
However, the false positive is really challenging to rely on this command.
[image: image.png]
[image: image.png]
[image: image.png]

Thanks,
Ben

**Comment by bmwluk2012** [comment]
Hi, I have done the suggested exercise today as listed below.
Backup-DbaDatabase -SqlInstance CARMD-IV-SQLMGT -Database MISBMM64
Backup the database 'MISBMM64' from carmd-iv-sql171 using the above command and restore the database 'MISBMM64' into carmd-iv-sqlmgt.  The two SQL instance are the same SQL version, then run the below command and got the output log as shown in the screenshot.  and I got the correct backup date value from CARMD-IV-SQLMGT, but still not getting the correct value in CARMD-IV-SQL171.  Do you see any reason why the date return are incorrect from CARMD-IV-SQL171?
Get-DbaLastBackup -SqlInstance carmd-iv-sql171 -Database MISBMM64 | Out-Gridview
Get-DbaLastBackup -SqlInstance carmd-iv-SQLMGT -Database MISBMM64 | Out-Gridview
![image](https://user-images.githubusercontent.com/36716515/90096408-06090800-dce8-11ea-9172-da705b42b2e1.png)
![image](https://user-images.githubusercontent.com/36716515/90096428-14efba80-dce8-11ea-8a70-e052d95d0d43.png)
![image](https://user-images.githubusercontent.com/36716515/90095900-d1488100-dce6-11ea-9b91-d482df4c2345.png)

**Comment by wsmelton** [comment]
Can you run the following command and verify if it returns the proper backup:

```
Get-DbaDbBackupHistory -SqlInstance CARMD-IV-SQLMGT -Database MISBMM64 -LastFull
Get-DbaDbBackupHistory -SqlInstance carmd-iv-SQLMGT -Database MISBMM64 -LastFull
```

**Comment by bmwluk2012** [comment]
I think you mean these 
Get-DbaDbBackupHistory -SqlInstance CARMD-IV-SQLMGT -Database MISBMM64 -LastFull | Out-Gridview
Get-DbaDbBackupHistory -SqlInstance carmd-iv-SQL171 -Database MISBMM64 -LastFull | Out-Gridview
But I still getting the incorrect date return from CARMD-IV-SQL171.
![image](https://user-images.githubusercontent.com/36716515/90100396-7405fd00-dcf1-11ea-901f-21eda3b382e0.png)

**Comment by andreasjordan** [comment]
And if you don't use -LastFull? So showing all full backups?

**Comment by bmwluk2012** [comment]
It will be like this as shown before the last.
![image](https://user-images.githubusercontent.com/36716515/90156328-29ac6c80-dd41-11ea-933d-9157a484d3e8.png)

**Comment by andreasjordan** [comment]
`Get-DbaLastBackup` only gets the last backup. 
`Get-DbaDbBackupHistory -LastFull` gets the last full backup.
What I want to see is the complete list of all backups, or better all full backups. So please use `Get-DbaDbBackupHistory -Type Full` to get these. There should be the one that is reported as the last but also others. I would like to see the reason for that backup beeing reported as last backup.

**Comment by bmwluk2012** [comment]
Yes, here are all the history on SQL171 regarding this database.
![image](https://user-images.githubusercontent.com/36716515/90158777-4ac28c80-dd44-11ea-83a0-7c7003c197f7.png)

**Comment by andreasjordan** [comment]
The wrong last backup is the first in the list, right?
Lets add the recovery fork and the LSN: place a `| Select-Object -Property Start, End, LastRecoveryForkGuid, FirstLsn` before the `| Out-GridView`. (Have no live system - there maybe typos in the command)
A lot of ordering is done with last_lsn. Maybe you have old records from an other database with the same name in the msdb, that had greater lsn. Then a way to solve the problem would be to clean these records from the backup-tables in msdb.

**Comment by bmwluk2012** [comment]
Yes, the incorrect date is the first record of the list.  I will try the suggested command first, thanks.

Per suggested "_A lot of ordering is done with last_lsn. Maybe you have old records from an other database with the same name in the msdb, that had greater lsn. Then a way to solve the problem would be to clean these records from the backup-tables in msdb._"
If you mean housekeeping the related record rows with database='MISBMM64'  from following tables in msdb resided in the SQL Instance carmd-iv-sql171.  Would that be any harm because it is a production SQL instance.
![image](https://user-images.githubusercontent.com/36716515/90174365-b6afef80-dd5a-11ea-8ba7-d56f9e7f8d93.png)

**Comment by bmwluk2012** [comment]
Here is the command I just run.
PS C:\> Get-DbaDbBackupHistory -SqlInstance carmd-iv-SQL171 -Database MISBMM64 -TYPE Full |
 Select-Object -Property SqlInstance, Database, Type, DeviceType, Start, Duration, End, LastRecoveryForkGuid, FirstLsn |
  Out-Gridview
![image](https://user-images.githubusercontent.com/36716515/90178905-8455c080-dd61-11ea-8f04-b6912843a443.png)
Questions:
1, Can I use 'order by' in the select statement in this commandlet
2. Frankly, my daily run command is actually using get-DbaLastbackup, not get-DbaDbBackupHistory, so the incorrect value is also returned from this command.

**Comment by andreasjordan** [comment]
It's 10 PM local time in germany, so I'll have a closer look in about 12 hours. But I think we can solve the case soon. Will answer your questions tomorrow.

**Comment by bmwluk2012** [comment]
Really appreciate for your help on the case Andreajordan.🙏

**Comment by andreasjordan** [comment]
First of all, there is a very good comment in the code:

https://github.com/sqlcollaborative/dbatools/blob/784387f8a066d69fb9c3dba92a3a0468858b4ff7/functions/Get-DbaDbBackupHistory.ps1#L474-L484

But it does not work as designed, I did a demo:

```
$SqlInstance = 'WIN10\SQL2016'
$DatabaseName = 'Test01'

$server = Connect-DbaInstance -SqlInstance $SqlInstance

$server.Databases['master'].ExecuteNonQuery("CREATE DATABASE $DatabaseName")
$database = $server.Databases[$DatabaseName]

$database.ExecuteNonQuery("CREATE TABLE dbo.test (x char(1000) default 'x')")
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName
1 .. 100 | ForEach-Object -Process { $database.ExecuteNonQuery("INSERT INTO dbo.test DEFAULT VALUES") }
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName

Get-DbaDbBackupHistory -SqlInstance $SqlInstance -Database $DatabaseName | Format-Table -Property End, LastRecoveryForkGuid, FirstLsn
<#
End                   LastRecoveryForkGUID                          FirstLsn
---                   --------------------                          --------
8/14/2020 12:21:58 PM 230d92b7-1a6e-4cda-a968-65bd6a63b63e 34000000107200042
8/14/2020 12:20:05 PM 230d92b7-1a6e-4cda-a968-65bd6a63b63e 34000000013600179
#>

$database | Remove-DbaDatabase -Confirm:$false

$server.Databases['master'].ExecuteNonQuery("CREATE DATABASE $DatabaseName")
$database = $server.Databases[$DatabaseName]

$database.ExecuteNonQuery("CREATE TABLE dbo.test (x char(1000) default 'x')")
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName
1 .. 100 | ForEach-Object -Process { $database.ExecuteNonQuery("INSERT INTO dbo.test DEFAULT VALUES") }
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName

Get-DbaDbBackupHistory -SqlInstance $SqlInstance -Database $DatabaseName | Format-Table -Property End, LastRecoveryForkGuid, FirstLsn
<#
End                   LastRecoveryForkGUID                          FirstLsn
---                   --------------------                          --------
8/14/2020 12:21:58 PM 230d92b7-1a6e-4cda-a968-65bd6a63b63e 34000000107200042
8/14/2020 12:24:18 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 34000000107200042
8/14/2020 12:20:05 PM 230d92b7-1a6e-4cda-a968-65bd6a63b63e 34000000013600179
8/14/2020 12:24:16 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 34000000013600179
#>

# I deleted the database (without deleting the historie) and created a new one with the same name.
# So the old backup history should not be included - have a look at the last part of the comment in the code.
# But it is - so there must be a bug there.

1 .. 1000 | ForEach-Object -Process { $database.ExecuteNonQuery("INSERT INTO dbo.test DEFAULT VALUES") }
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName
1 .. 1000 | ForEach-Object -Process { $database.ExecuteNonQuery("INSERT INTO dbo.test DEFAULT VALUES") }
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName
1 .. 1000 | ForEach-Object -Process { $database.ExecuteNonQuery("INSERT INTO dbo.test DEFAULT VALUES") }
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName
Get-DbaDbBackupHistory -SqlInstance $SqlInstance -Database $DatabaseName | Format-Table -Property End, LastRecoveryForkGuid, FirstLsn
<#
End                   LastRecoveryForkGUID                          FirstLsn
---                   --------------------                          --------
8/14/2020 12:26:43 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 38000000900000042
8/14/2020 12:26:41 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 38000000092800042
8/14/2020 12:26:39 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 36000000124000042
8/14/2020 12:21:58 PM 230d92b7-1a6e-4cda-a968-65bd6a63b63e 34000000107200042
8/14/2020 12:24:18 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 34000000107200042
8/14/2020 12:20:05 PM 230d92b7-1a6e-4cda-a968-65bd6a63b63e 34000000013600179
8/14/2020 12:24:16 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 34000000013600179
#>

1 .. 1000 | ForEach-Object -Process { $database.ExecuteNonQuery("INSERT INTO dbo.test DEFAULT VALUES") }
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName
1 .. 1000 | ForEach-Object -Process { $database.ExecuteNonQuery("INSERT INTO dbo.test DEFAULT VALUES") }
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName
1 .. 1000 | ForEach-Object -Process { $database.ExecuteNonQuery("INSERT INTO dbo.test DEFAULT VALUES") }
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName
Get-DbaDbBackupHistory -SqlInstance $SqlInstance -Database $DatabaseName | Format-Table -Property End, LastRecoveryForkGuid, FirstLsn
<#
End                   LastRecoveryForkGUID                          FirstLsn
---                   --------------------                          --------
8/14/2020 12:27:56 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 39000000047200042
8/14/2020 12:27:54 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 38000002515200042
8/14/2020 12:27:52 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 38000001707200042
8/14/2020 12:26:43 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 38000000900000042
8/14/2020 12:26:41 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 38000000092800042
8/14/2020 12:26:39 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 36000000124000042
8/14/2020 12:21:58 PM 230d92b7-1a6e-4cda-a968-65bd6a63b63e 34000000107200042
8/14/2020 12:24:18 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 34000000107200042
8/14/2020 12:20:05 PM 230d92b7-1a6e-4cda-a968-65bd6a63b63e 34000000013600179
8/14/2020 12:24:16 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 34000000013600179
#>

# Now I restore the database to an earlier backup to get a new recovery fork
Get-DbaDbBackupHistory -SqlInstance $SqlInstance -Database $DatabaseName | Where-Object FirstLsn -EQ '36000000124000042' | Restore-DbaDatabase -SqlInstance $SqlInstance -WithReplace

# Lets take a backup and have a look
$null = Backup-DbaDatabase -SqlInstance $SqlInstance -Database $DatabaseName
Get-DbaDbBackupHistory -SqlInstance $SqlInstance -Database $DatabaseName | Format-Table -Property End, LastRecoveryForkGuid, FirstLsn
<#
End                   LastRecoveryForkGUID                          FirstLsn
---                   --------------------                          --------
8/14/2020 12:27:56 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 39000000047200042
8/14/2020 12:27:54 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 38000002515200042
8/14/2020 12:27:52 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 38000001707200042
8/14/2020 12:26:43 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 38000000900000042
8/14/2020 12:26:41 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 38000000092800042
8/14/2020 12:33:28 PM 46920f9e-367c-4e7f-91ac-e3592b225f7a 36000000153600081  <==
8/14/2020 12:26:39 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 36000000124000042
8/14/2020 12:21:58 PM 230d92b7-1a6e-4cda-a968-65bd6a63b63e 34000000107200042
8/14/2020 12:24:18 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 34000000107200042
8/14/2020 12:20:05 PM 230d92b7-1a6e-4cda-a968-65bd6a63b63e 34000000013600179
8/14/2020 12:24:16 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 34000000013600179
#>

# I marked the new backup with <==
# Because the backups are sorted by FirstLsn, my last backup is not the "LastBackup"
Get-DbaDbBackupHistory -SqlInstance $SqlInstance -Database $DatabaseName -LastFull | Format-Table -Property End, LastRecoveryForkGuid, FirstLsn
<#
End                   LastRecoveryForkGUID                          FirstLsn
---                   --------------------                          --------
8/14/2020 12:27:56 PM ec32997b-c70b-4bf4-affd-014aef1c7b00 39000000047200042
#>
```

First of all, how can @bmwluk2012 solve the problem:

You could just delete all entries in the backup tables in msdb not related to your current recovery fork. The data in these tables are only there to help a restore on that instance. But there a constraints, so it's not a one liner. 
But you can change the database name in the backupset table:
```
UPDATE msdb..backupset SET database_name = database_name + '_invisible' WHERE database_name = 'Test01' and last_recovery_fork_guid != '46920F9E-367C-4E7F-91AC-E3592B225F7A'
```
The data is still there, so no record deleted in a production instance, but kind of marked "invisible".

And now to the moriginal editors ( @potatoqualitee / @Stuart-Moore ): Would you like me to rewrite the sql to address this issue? Should we open a new issue for that or use this one? Or am I missing something and this is no bug?

**Comment by wsmelton** [comment]
What is your proposed change to the code to address the issue?

**Comment by andreasjordan** [comment]
Because I like SQL I played a little bit with the query.

I think we can change these lines:
https://github.com/sqlcollaborative/dbatools/blob/784387f8a066d69fb9c3dba92a3a0468858b4ff7/functions/Get-DbaDbBackupHistory.ps1#L566-L575

to this:
```
                        JOIN (
                        SELECT database_guid, last_recovery_fork_guid
                        FROM (SELECT database_guid, last_recovery_fork_guid, ROW_NUMBER() OVER (ORDER BY backup_finish_date DESC) AS rn
                              FROM msdb..backupset
                              WHERE database_name = '$($db.Name)'
                             ) AS bs
                        WHERE rn = 1
                        ) AS last_guids ON last_guids.database_guid = backupset.database_guid AND last_guids.last_recovery_fork_guid = backupset.last_recovery_fork_guid
                    WHERE 1=1 $whereCopyOnly

```

That would get the last database_guid (is maybe not needed if last_recovery_fork_guid is unique over all databases) and the last used last_recovery_fork_guid for the database to filter the possible last backups. Then they can savely sorted by LastLSN.

**Comment by wsmelton** [comment]
@Stuart-Moore authored and owns our backup/restore process so he gets the review and final call on this one.

**Comment by bmwluk2012** [comment]
> What is your proposed change to the code to address the issue?

Regarding the suggested workaround, I have tried and it works perfectly now, no more false positive value returns, see my actions as shown following
Run the TSQLs ...
use msdb;
/*Identified the related/required last_recovery_fork_guid,*/
--CARMD-IV-SQL171	MISBMM64	Full	Disk	8/12/2020 9:45:26 PM	00:00:01	8/12/2020 9:45:27 PM	1cc9f673-d6d4-437b-9220-e374079ff33a	92000000046300037	
/*Action*/
Begin Tran T1
	update backupset
		set  database_name = database_name + '_invisible'
		 WHERE last_recovery_fork_guid<>'1cc9f673-d6d4-437b-9220-e374079ff33a' and database_name = 'MISBMM64'
--(1904 rows affected);   Completion time: 2020-08-14T08:23:43.1447451-07:00
Commit Tran T1	--Completion time: 2020-08-14T08:25:44.0602198-07:00
SELECT * from backupset WHERE last_recovery_fork_guid<>'1cc9f673-d6d4-437b-9220-e374079ff33a' and database_name = 'MISBMM64_invisible'	--1904, not the latest backup set

Then run the DBATool commandlet as ....
PS C:\> get-DbaLastBackup  -SqlInstance  carmd-iv-sql171 -Database MISBMM64 | Select * | Out-Gridview
And the result set has return as expected ...
![image](https://user-images.githubusercontent.com/36716515/90267377-4a8cc480-de0a-11ea-8900-0a24d4108495.png)
Thanks for the help @andreasjordan 🙏

**Comment by niphlod** [comment]
gosh I love the fact that I was the one commenting on that query or I'd have an headache trying to reconstruct why we do that. 
Theoretically it seems that we wanted to prevent a problem that was fixed but not on all cornercases. Though I'm still unsure why the current query (which doesn't consider fork_guid) doesn't solve the problem at all. It SEEMS to select correctly the database_guid which has the most recent finish-date.
My notes from two years ago are only a minimal corner-case where DST happens (or timezone in the server is set to a different one) and you created a new db in between ...

Query-wise, ROW_NUMBER() on 2005 surely works (do not ask where, though). 
But I'll try nonetheless to reproduce the error and spot where the current query breaks

**Comment by andreasjordan** [comment]
@niphlod The database_guid only helps to filter out different databases with the same name. When you have a database with a very high LSN and you have to restore a backup with a low LSN (so going way back in history - why that ever is usefull - but was done here in this case in this issue) the database_guid stays the same, only the recovery_fork_guid changes and the LSNs are back to the low value. Then sorting just by LSN (which is very good and the reason for that is very good documentated in the code - thanks for that!) gets the backup with the high LSN from the old recovery fork that is not active at the moment. So I introduced the filter to the current recovery_fork. If you want to replay it, have a look at my demo: https://github.com/sqlcollaborative/dbatools/issues/6730#issuecomment-674019765

**Comment by niphlod** [comment]
@andreasjordan : aaargh, gotcha. 
Note to self (to read in about 2 years lol): I remember now why we added that part. 
It was when you restore dbs with the same name over and over BETWEEN environments, i.e. a refresh from prod to test. In that case database_guid is different hence it works. Here instead we're restoring in-place and it bombs, given database_guid doesn't change.

The snippet was incredibly useful. But .... we can totally avoid ROW_NUMBER() (and IMHO get a more legible query) with a rather simple

```
JOIN msdb..backupset AS backupset ON backupset.media_set_id = mediaset.media_set_id
                        JOIN (
                            SELECT TOP 1 database_guid, last_recovery_fork_guid, backup_finish_date
                            FROM msdb..backupset
                            WHERE database_name = '$($db.Name)'
                            ORDER BY backup_finish_date DESC
                        ) AS last_guids ON last_guids.database_guid = backupset.database_guid AND last_guids.last_recovery_fork_guid = backupset.last_recovery_fork_guid
                    WHERE backupset.database_name = '$($db.Name)' $whereCopyOnly
```

**Comment by andreasjordan** [comment]
@niphlod Thanks for reminding me that we are on SQL Server and not on Oracle. Most SQL stuff I do, I do on Oracle databases so I often forget the nice TOP 1 feature from Microsoft. I will update the pull request first thing tomorrow morning - and now go back to my friends and my glas of wine.

**Comment by niphlod** [comment]
@andreasjordan : noprobls. Maybe add a comma to that comment, I feel like it'll still help down the line to retrace the steps we took (and the thinking behind).
if you don't want to add a regression test to Get-DbaDbBackupHistory.Tests.ps1 I'll be happy to punch that in reusing your snippet.

to all who partecipated in this issue: thanks for the patience and the hints (and pushing for a real-case scenario even if I think I proved that remembering things is a hard job )
